File: sento-sample/src/main/java/io/sento/sample/annotations/OnFocusLost.java
package io.sento.sample.annotations;

import io.sento.ListenerBinding;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Retention(RetentionPolicy.CLASS)
@Target(ElementType.METHOD)
@ListenerBinding(
    owner = "android.view.View",
    listener = "io.sento.sample.annotations.OnFocusLostListener",
    setter = "setOnFocusChangeListener"
)
public @interface OnFocusLost {
  public int[] value();
}

File: sento-sample/src/main/java/io/sento/sample/annotations/OnFocusLost.kt
package io.sento.sample.annotations

import android.view.View
import io.sento.annotations.ListenerClass

@Retention(AnnotationRetention.RUNTIME)
@Target(AnnotationTarget.FUNCTION, AnnotationTarget.PROPERTY_GETTER, AnnotationTarget.PROPERTY_SETTER)
@ListenerClass(
    owner = "android.view.View",
    listener = "io.sento.sample.annotations.OnFocusLostListener",
    setter = "setOnFocusChangeListener",
    callback = "onFocusLost"
)
annotation class OnFocusLost(vararg val value: Int)

abstract class OnFocusLostListener : View.OnFocusChangeListener {
  override fun onFocusChange(view: View, focused: Boolean) {
    if (!focused) {
      onFocusLost(view)
    }
  }

  abstract fun onFocusLost(view: View)
}
